### PR TITLE
Implement Dev Services for Kafka

### PR DESCRIPTION
**Pull request by cescoffier**
Add support for Kafka as Dev Service.

It uses Red Panda as it's way faster (more than 20 seconds using the official image, ~ 15 seconds with Strimzi, less than 1 second with Red Panda).

### CODE REVIEW

File: docs/src/main/asciidoc/kafka.adoc [hunk 1]
```diff
@@ -338,15 +302,66 @@ Then, you just need to run the application using:
...
+You can build and run the application in JVM mode with:
+
+[source, bash]
+----
+./mvnw package
+java -jar target/quarkus-app/quarkus-run.jar
+----
 
-== Running Native
+NOTE: By default, the application tries to connect to a Kafka broker listening to `localhost:9092`.
```
Comment (Ladicek) [inline docs/src/main/asciidoc/kafka.adoc:356]:
```suggestion
NOTE: By default, the application tries to connect to a Kafka broker listening at `localhost:9092`.
```

File: docs/src/main/asciidoc/kafka.adoc [hunk 2]
```diff
@@ -338,15 +302,66 @@ Then, you just need to run the application using:
...
+You can configure the bootstrap server using: `java -Dkafka.bootstrap.servers=... -jar target/quarkus-app/quarkus-run.jar`
 
-You can build the native executable with:
+You can build and the native executable with:
```
Comment (Ladicek) [inline docs/src/main/asciidoc/kafka.adoc:359]:
```suggestion
You can build and run the native executable with:
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            super(dockerImageName);
+            this.port = fixedExposedPort;
+            withNetwork(Network.SHARED);
+            withExposedPorts(KAFKA_PORT);
+            // For red panda, we need to start the broker - see https://vectorized.io/docs/quick-start-docker/
+            if (dockerImageName.getRepository().equals("vectorized/redpanda")) {
+                withCreateContainerCmdModifier(cmd -> {
+                    cmd.withEntrypoint("sh");
+                });
+                withCommand("-c", "while [ ! -f " + STARTER_SCRIPT + " ]; do sleep 0.1; done; " + STARTER_SCRIPT);
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:249]: Given the starter script is just a single command, wouldn't it be better to just define it here? Something like:

```java
withCommand("-c", "/usr/bin/rpk redpanda start ...");
```

?

(Perhaps it doesn't even have to be executed in shell, but I don't really know.)

Comment (Ladicek) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:249]: Ah sorry, stupid question. At this point, you don't know the mapped port. Please ignore :-)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                    Runtime.getRuntime().removeShutdownHook(closeHookThread);
+                }
+            });
+        }
+        cfg = configuration;
+        return bootstrapServers;
+    }
+
+    private void shutdownBroker() {
+        if (closeable == null) {
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:104]:
```suggestion
        if (closeable != null) {
```

???

Otherwise NPE is guaranteed.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+    private KafkaBroker startKafka(KafkaDevServiceCfg config) {
+        if (!config.devServicesEnabled) {
+            // explicitly disabled
+            log.debug("Not starting dev services for Kafka, as it has been disabled in the config.");
+            return null;
+        }
+
+        if (!isDockerWorking.getAsBoolean()) {
+            log.warn("Please configure the Kafka bootstrap servers property (kafka.bootstrap.servers) "
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:123]: I'd rephrase this, it's not exactly clear what the problem is. Something like `Docker isn't working, please configure the Kafka bootstrap servers property (kafka.bootstrap.servers)`.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+            log.debug("Not starting dev services for Kafka, the kafka.bootstrap.servers is configured.");
+            return null;
+        }
+
+        // Verify that we have kafka channels without bootstrap.servers
+        if (!hasKafkaChannelWithoutBootstrapServers()) {
+            log.debug("Not starting dev services for Kafka, all the channels are configured.");
+            return null;
+        }
+
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:139]: I don't know anything about our Kafka Streams extension, but maybe that should also be checked?

**Comment by Ladicek** [comment]
I don't know anything about Dev Services (particularly how state is supposed to be managed in the processor class), but other than that, this looks OK baring a few comments.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 3 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:249]: I need to try, I reused the approach from Test Container.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 6 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:139]: it uses `kafka.bootstrap.servers` - so we're covered.

Comment (Ladicek) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:139]: Great!

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 3 above)
Comment (Ladicek) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:249]: Nah, it needs to be like this, I asked before thinking :-)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 4 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:104]: and now the question: why is this not called?

File: docs/src/main/asciidoc/kafka-dev-services.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+= Dev Services for Kafka
```
Comment (gsmet) [inline docs/src/main/asciidoc/kafka-dev-services.adoc:6]: Does it really make sense to have a specific document just for that?

Not sure how it's organized for others but it feels a bit weird.

Comment (cescoffier) [reply docs/src/main/asciidoc/kafka-dev-services.adoc:6]: I don't want to have it in the Kafka quickstart (which has a lot of extra content that needs to move). Dev Services and its configuration has nothing to do with the quickstart. When we will have a proper reference guide it will go there. 

Having a Kafka reference guide is on my todo list.

**Comment by geoand** [comment]
Seems very similar to what we have for Mongo and Redis, so LGTM

**Comment by sberyozkin** [comment]
Should there be some template support for it as well ? As in the `quarkus-cache` for example ?

**Comment by sberyozkin** [comment]
It can be complemented later for sure if it can be of interest

**Comment by pjgg** [comment]
Kafka DevServices! Great news!!!. 

Just a quick one, does this first release of Kafka DevServices support AVRO (a schema registry)?

**Comment by cescoffier** [comment]
@pjgg it supports Avro, but would not start a schema registry. We would need to find a slim, fast to start schema registry first.

**Comment by cescoffier** [comment]
@sberyozkin @geoand it looks similar but is not exactly the same. Kafka requires customizing the container to configure the advertised brokers. So, the docker/container part is slightly more convoluted (well, not the end of the world).

**Comment by Ladicek** [comment]
Unrelated to this PR, but -- when we add extension(s) for Apicurio Registry (the Avro library will be first, but Protobuf will likely follow), we can add Dev Services support for schema registry. The Apicurio Registry is pretty fast to start (it's a Quarkus app after all! :-) ), so should be fine IMHO.

**Comment by cescoffier** [comment]
@Ladicek ideally, we would have a "slim" container image that we can just run.

**Comment by Ladicek** [comment]
Apicurio Registry has a container image with in-memory storage; I think we can easily use that: https://hub.docker.com/r/apicurio/apicurio-registry-mem/tags

**Comment by cescoffier** [comment]
Awesome! Exactly what we need.

**Comment by cescoffier** [comment]
Does it needs to access the Kafka broker (the dev services for kafka emits a build item with the bootstrap server, so we should be fine, but we will have to check)

**Comment by Ladicek** [comment]
It doesn't need to access the Kafka broker -- not in the in-memory variant. They have Kafka-based storage as well, but I assume we wouldn't use that for dev services.

**Comment by rsvoboda** [comment]
Red Panda was picked for Dev Services. What if user wants to use the official image or Strimzi? Is the Kafka DevServices going to be customizable to allow Strimzi for example?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    @ConfigItem
+    public Optional<Integer> port;
+
+    /**
+     * The Kafka image to use.
+     * Note that only Red Panda images are supported.
+     * See https://vectorized.io/docs/quick-start-docker/ and https://hub.docker.com/r/vectorized/redpanda
+     */
+    @ConfigItem(defaultValue = "vectorized/redpanda:latest")
```
Comment (gsmet) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:33]: I'm surprised you went with `latest`? I thought you were against this?

Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:33]: Still puzzled about that actually. Let me change that.

**Comment by cescoffier** [comment]
@rsvoboda at the moment it only supports red panda. 

We can add strimzi, but the experience was quite bad on my machine. Note that it's not just changing the image name, the whole startup script, kafka advertising, and so on need to be done for that one too. Hopefully, we would get images using Kraft. 

Don't forget that the user can still start its own cluster (strimzi or whatever he likes) and disables the dev services.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java (hunk 8 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java]: fixed.

**Review by stuartwdouglas — APPROVED** [review]
LGTM, but I just realized that we have actually missed native support in this and some of the other dev services PRs. It's probably best to just merge this one and then fix them all together.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
Actually looks like it was added for the other ones, but is missing for this one, see io.quarkus.test.junit.NativeDevServicesHandler.

TBH I don't like that these are separate build items, they should be refactored to a single common one, it really just sets system properties so this could be simplified a lot I think.

**Comment by stuartwdouglas** [comment]
https://github.com/quarkusio/quarkus/pull/17574 is what I am talking about in terms of a generic build item.

**Comment by cescoffier** [comment]
So basically producing one per "key"?

**Comment by cescoffier** [comment]
BTW, Trying to rush the integration of NativeDevServicesHandler. I need this in 2.0 final.

**Comment by cescoffier** [comment]
I've added the DevServicesNativeConfigResultBuildItem.

**Comment by cescoffier** [comment]
@stuartwdouglas can you have a look?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        if (!isDockerWorking.getAsBoolean()) {
+            log.warn("Docker isn't working, please configure the Kafka bootstrap servers property (kafka.bootstrap.servers).");
+            return null;
+        }
+
+        // Check if kafka.bootstrap.servers is set
+        if (ConfigUtils.isPropertyPresent(KAFKA_BOOTSTRAP_SERVERS)) {
```
Comment (stuartwdouglas) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:141]: Shouldn't this check be above the docker one? Otherwise if you have configured `kafka.bootstrap.servers` but don't have docker you will get a weird warning telling you to configure `kafka.bootstrap.servers` even though you already have.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+            log.debug("Not starting dev services for Kafka, the kafka.bootstrap.servers is configured.");
+            return null;
+        }
+
+        // Verify that we have kafka channels without bootstrap.servers
+        if (!hasKafkaChannelWithoutBootstrapServers()) {
```
Comment (stuartwdouglas) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:147]: Same here.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

**Review by stuartwdouglas — APPROVED** [review]
